TTL fails with `features that prevent cachability are disabled on this context`

Start with 2.4.0:

```
PUT my_index
{
  "mappings": {
    "my_type": {
      "_ttl": {
        "enabled": true
      }
    }
  }
}
```

Upgrade to 5.0.0:

```
PUT my_index/my_type/1?ttl=1s
{
  "text": "Will expire in 10 minutes"
}


GET my_index/_search
```

Returns:

```
{
  "took": 34,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 4,
    "failed": 1,
    "failures": [
      {
        "shard": 3,
        "index": "my_index",
        "node": "AFATqx6CR6SPvGWQ0DaxvA",
        "reason": {
          "type": "illegal_argument_exception",
          "reason": "features that prevent cachability are disabled on this context"
        }
      }
    ]
  },
  "hits": {
    "total": 1,
    "max_score": 1,
    "hits": []
  }
} 
```

Related to https://github.com/elastic/elasticsearch/pull/21303

Stacktrace:

```
[2016-11-10T09:42:48,016][DEBUG][o.e.a.s.TransportSearchAction] [AFATqx6] [3] Failed to execute fetch phase
org.elasticsearch.transport.RemoteTransportException: [AFATqx6][127.0.0.1:9300][indices:data/read/search[phase/fetch/id]]
Caused by: java.lang.IllegalArgumentException: features that prevent cachability are disabled on this context
       	at org.elasticsearch.index.query.QueryShardContext.failIfFrozen(QueryShardContext.java:396) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.index.query.QueryShardContext.nowInMillis(QueryShardContext.java:278) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.index.mapper.TTLFieldMapper$TTLFieldType.valueForSearch(TTLFieldMapper.java:146) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.index.fieldvisitor.FieldsVisitor.postProcess(FieldsVisitor.java:93) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.search.fetch.FetchPhase.createSearchHit(FetchPhase.java:187) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:152) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:474) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.action.search.SearchTransportService.lambda$registerRequestHandler$684(SearchTransportService.java:311) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.transport.TransportService$6.doRun(TransportService.java:548) [elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:520) [elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.0.1-SNAPSHOT.jar:5.0.1-SNAPSHOT]
       	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_73]
       	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_73]
       	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_73]
```

--- a/core/src/main/java/org/elasticsearch/index/mapper/TTLFieldMapper.java
+++ b/core/src/main/java/org/elasticsearch/index/mapper/TTLFieldMapper.java
@@ -28,8 +28,6 @@ import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentParser;
 import org.elasticsearch.index.AlreadyExpiredException;
-import org.elasticsearch.search.internal.SearchContext;
-
 import java.io.IOException;
 import java.util.Date;
 import java.util.Iterator;
@@ -140,13 +138,7 @@ public class TTLFieldMapper extends MetadataFieldMapper {
         // Overrides valueForSearch to display live value of remaining ttl
         @Override
         public Object valueForSearch(Object value) {
-            long now;
-            SearchContext searchContext = SearchContext.current();
-            if (searchContext != null) {
-                now = searchContext.getQueryShardContext().nowInMillis();
-            } else {
-                now = System.currentTimeMillis();
-            }
+            long now = System.currentTimeMillis();
             Long val = (Long) super.valueForSearch(value);
             return val - now;
         }

--- a/core/src/test/java/org/elasticsearch/ttl/SimpleTTLIT.java
+++ b/core/src/test/java/org/elasticsearch/ttl/SimpleTTLIT.java
@@ -26,6 +26,7 @@ import org.elasticsearch.action.admin.indices.mapping.put.PutMappingResponse;
 import org.elasticsearch.action.admin.indices.stats.IndicesStatsResponse;
 import org.elasticsearch.action.get.GetResponse;
 import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.action.search.SearchResponse;
 import org.elasticsearch.action.update.UpdateRequestBuilder;
 import org.elasticsearch.action.update.UpdateResponse;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
@@ -48,6 +49,7 @@ import java.util.concurrent.TimeUnit;
 import static org.elasticsearch.action.support.WriteRequest.RefreshPolicy.IMMEDIATE;
 import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertSearchResponse;
 import static org.hamcrest.Matchers.both;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
@@ -301,5 +303,26 @@ public class SimpleTTLIT extends ESIntegTestCase {
         String ttlAsString = mappingSource.get("_ttl").toString();
         assertThat(ttlAsString, is(notNullValue()));
         assertThat(errMsg, ttlAsString, is("{enabled=true}"));
+    }
+
+    // Test for #21457
+    public void testSearchWithTTL() throws Exception {
+        assertAcked(prepareCreate("test")
+                .setSettings(IndexMetaData.SETTING_VERSION_CREATED, Version.V_2_3_0.id)
+                .addMapping("type1", XContentFactory.jsonBuilder()
+                        .startObject()
+                        .startObject("type1")
+                        .startObject("_ttl").field("enabled", true).endObject()
+                        .endObject()
+                        .endObject()));
+
+        long providedTTLValue = 300000;
+        IndexResponse indexResponse = client().prepareIndex("test", "type1", "1").setSource("field1", "value1")
+                .setTTL(providedTTLValue).setRefreshPolicy(IMMEDIATE).get();
+        assertEquals(DocWriteResponse.Result.CREATED, indexResponse.getResult());
+
+        SearchResponse searchResponse = client().prepareSearch("test").get();
+        assertSearchResponse(searchResponse);
+        assertEquals(1L, searchResponse.getHits().getTotalHits());
     }
 }

